### ISSUE TITLE
Host coqdoc documentation online

### ISSUE THREAD

**Issue by langston-barrett** [UniMath/UniMath#805]
This isn't urgent, but would be great to have in the future. The HoTT repo has an _incredible_ Travis CI setup, and might be a good place to look for how this is done (they even use Privola to allow people to step through proofs in the browser). 

More generally, this documentation should be proofread and vastly expanded, there's a lot of uncommented code. Even if it seems obvious when you write it, it saves people time when reading the code to have it "decoded" into natural language in a comment.

**Comment by abooij** [comment]
I'd be happy to look into setting this up. I don't know privola but I can see if that's easy to set up as well. I think it would be useful to have this for next week.

**Comment by benediktahrens** [comment]
On 12/07/2017 11:52 PM, Langston Barrett wrote:
> This isn't urgent, but would be great to have in the future. The HoTT
> repo has an _incredible_ Travis CI setup, and might be a good place
> to look for how this is done (they even use Privola to allow people
> to step through proofs in the browser).
> 
> More generally, this documentation should be proofread and vastly
> expanded, there's a lot of uncommented code. Even if it seems obvious
> when you write it, it saves people time when reading the code to have
> it "decoded" into natural language in a comment.
> 


For a start, I could reactivate the build script I had running nightly 
on an IAS server. It put the coqdoc here:

http://www.math.ias.edu/~ahrens/UniMath/enhanced-html/


However, it would be better to have something done by Travis, if that is 
possible.

**Comment by siddharthist** [comment]
@benediktahrens That seems like a good intermediate step! I think Travis is the right tool for this job. Every Github repo comes with its own domain name and ability to serve static HTML (this is a feature called Github Pages), which I think is a good place for Travis to send the coqdoc.

**Comment by abooij** [comment]
I have an attempt at making this work from travis in my branch [travis-coqdoc](https://github.com/abooij/UniMath/tree/travis-coqdoc). At the end of every CI test on a given branch, it compiles the html documentation, and commits it to the `gh-pages` branch as the only commit. So it does not leave a long history of previous versions of the website, as this is generated anyway, and so the history is not interesting.

The documentation deployment is the last step of the CI. By marking the deployment step as an "allowed failure" with "fast finishing", if the actual test stages have finished, then everything lights up green, and it looks like the build is done and passed, even though it may still be working on the documentation. Building the documentation takes rather long as it does not cache the build results from the test stages.

To push a commit, travis needs some kind of push access. The solution I've chosen here is to make use of github's "deploy keys", which are like SSH keypairs for a specific repository. I have set it up for my fork, but it can also be set up for the UniMath fork of UniMath.

See the CI and html build & upload [here](https://travis-ci.org/abooij/UniMath/builds/313760371). The deployed docs are [here](https://abooij.github.io/UniMath/).

**Comment by DanGrayson** [comment]
We build individual parts of UniMath in travis sub-jobs, so how would we generate the documentation for all of UniMath, practically speaking?

**Comment by abooij** [comment]
@DanGrayson, what do you mean? The final documentation deployment job that I added builds all of UniMath. This takes between 30 and 45 minutes.

**Comment by DanGrayson** [comment]
I think the time limit for travis is 45 minutes, so as UniMath grows, the first travis job to bump into the time limit will be this new documentation job.  (Currently all jobs run in less than twenty minutes, and all but one run in less than 10 minutes.)

**Comment by abooij** [comment]
@DanGrayson, I see your point. So at some point this may stop automatically updating the documentation.

It is possible to combine the build products from the other stages, so that this documentation build can be much faster. However, that is more difficult to set up, as it requires uploading those files at the end of the test stages, for instance to an S3 instance. So at the moment I haven't done that.

**Comment by abooij** [comment]
(By the way, I believe the time limit is 50 minutes.)

**Comment by DanGrayson** [comment]
Adding this travis job would cause our travis check to fail sooner that it normally would, as UniMath grows, so I think we should not do it this way.

**Comment by abooij** [comment]
It wouldn't make any checks fail, since this deployment job is marked as "allowed failure" (see also my long comment above). In the worst case, at some point it will stop updating the online documentation, or not manage to do it always (as build time is a bit erratic).

**Comment by DanGrayson** [comment]
Does it detect which job is the one allowed to fail by comparison of environment variable settings?
```
  allow_failures:
    - env: BUILD_ALSO="html" DEPLOY=yes
```

**Comment by abooij** [comment]
[Yes](https://docs.travis-ci.com/user/customizing-the-build/#Matching-Jobs-with-allow_failures).

**Comment by arnoudvanderleer** [comment]
@DanGrayson @benediktahrens Is this still something that we would want? Because I think it would be 30 or maybe 60 minutes of work to copy the relevant parts of the [HoTT CI](https://github.com/HoTT/Coq-HoTT/blob/master/.github/workflows/ci.yml) and have our CI build the documentation and publish it with Jekyll. Personally, I think this would be much better than [the very irregular, manual pushing we currently do at UniMath.github.io](https://github.com/UniMath/unimath.github.io/tree/master/doc/UniMath).

**Comment by benediktahrens** [comment]
@arnoudvanderleer : each version of the coqdoc is many MB big. I wonder how quickly a repository hosting the documentation for each merged PR to UniMath would grow - maybe it would become unfeasibly large quickly?

**Comment by nmvdw** [comment]
> @arnoudvanderleer : each version of the coqdoc is many MB big. I wonder how quickly a repository hosting the documentation for each merged PR to UniMath would grow - maybe it would become unfeasibly large quickly?

To give some data:
- The total html documentation is 580 MB.

**Comment by arnoudvanderleer** [comment]
@benediktahrens I believe the HoTT repository does not store every version (why do we store so many different documentation versions in our GitHub repo anyway?).
Their CI merely builds the new documentation and then displays that one online via Jekyll. I think that would be a good addition, since our current online documentation is 2 years old.

**Comment by arnoudvanderleer** [comment]
And even if we'd store all the versions, note that the coqdoq is just an inflated version of the code. Since git does some compression, I hope/suspect that the overhead won't be too large. And then the documentation and code repos would grow at similar rates.

**Comment by benediktahrens** [comment]
> @benediktahrens I believe the HoTT repository does not store every version (why do we store so many different documentation versions in our GitHub repo anyway?).

It is important here to distinguish two things: 
1. the showing of the documentation of the latest version of UniMath
2. the long-term archiving of specific versions of the documentation for referencing in publications

I think that these two are confused here:

> Personally, I think this would be much better than [the very irregular, manual pushing we currently do at UniMath.github.io](https://github.com/UniMath/unimath.github.io/tree/master/doc/UniMath).

@arnoudvanderleer : Are you suggesting we do something about 1.?

**Comment by arnoudvanderleer** [comment]
Do we currently show the latest version of UniMath somewhere?

**Comment by benediktahrens** [comment]
No, we do not; it would be nice to do it.

**Comment by nmvdw** [comment]
> No, we do not; it would be nice to do it.

The suggestion by Arnoud would give a way to display the documentation of the latest version of UniMath. It would not help with keeping track of the documentation of older versions used in papers (even if we restrict ourselves to tags), because then the storage might get out of hand.  For that reason, I do think it would be a beneficial change, as long as we acknowledge that this is only for the latest version.

**Comment by arnoudvanderleer** [comment]
Yes, I realize now that the repository with the only HTML documentation that we currently have online, is only for archiving purposes. All the more reason then to implement hosting this documentation.

**Comment by arnoudvanderleer** [comment]
This was completed by #2032
